python/nav/ipdevpoll/shadows/__init__.py

```diff
@@ -259,6 +266,38 @@ def _fix_binary_garbage(self):
                 setattr(self, attr, repr(value))
         self.clear_cached_objects()
 
+    def _detect_version_changes(self):
+        """
+        Detects if the software, hardware or firmware version changed for each device.
+
+        Saves this information in changed_versions in the Device instance.
+        """
+        old_device = self.get_existing_model()
+        self.changed_versions = ()
+        if old_device:
+            self.changed_versions = set(self.get_diff_attrs(old_device)).intersection(
+                (
+                    'hardware_version',
+                    'software_version',
+                    'firmware_version',
+                )
+            )
+
+    def cleanup(self, containers):
+        if self.changed_versions:
+            self._post_events_version_changes()
+
+    def _post_events_version_changes(self):
+        """Posts events for software, hardware or firmware changes."""
+        event = EventFactory('ipdevpoll', 'eventEngine', 'deviceNotice')
```

Comment:
The typical usage pattern for `EventFactory` is that you create the factory instance only *once*, e.g. as a class variable (or even, a global variable, if need be). Your code creates a new factory every time you need a new event.